app/services/product_service.py

```diff
@@ -0,0 +1,129 @@
+            )
+
+        new_product = Product(**product.dict())
+        self.db.add(new_product)
+        self.db.commit()
+        self.db.refresh(new_product)
+        return new_product
+
+    def update_product(
+        self, product_id: UUID, product_update: UpdatedProductRequestModel
+    ) -> Product:
+        product = self.db.query(Product).filter(Product.id == product_id).first()
+        if not product:
+            raise HTTPException(
+                status_code=status.HTTP_404_NOT_FOUND, detail="Product not found"
+            )
+
+        update_data = product_update.dict(exclude_unset=True)
+
+        # Check for unique name if it's being updated
+        if "name" in update_data:
+            existing_product = (
+                self.db.query(Product)
+                .filter(
+                    Product.name.ilike(update_data["name"]), Product.id != product_id
+                )
+                .first()
+            )
+            if existing_product:
+                raise HTTPException(
+                    status_code=status.HTTP_400_BAD_REQUEST,
+                    detail=f"Product name '{update_data['name']}' already exists. Please use a unique name.",
+                )
+
+        for key, value in update_data.items():
+            setattr(product, key, value)
+
+        product.updated_at = datetime.now(timezone.utc)
+        self.db.commit()
+        self.db.refresh(product)
+        return product
+
+    def delete_product(self, product_id: UUID):
+        product = self.db.query(Product).filter(Product.id == product_id).first()
+        if not product:
+            raise HTTPException(
+                status_code=status.HTTP_404_NOT_FOUND,
+                detail=f"Product with ID {product_id} not found.",
+            )
+        self.db.delete(product)
+        self.db.commit()
+
+    def get_all_products(self):
+        return self.db.query(Product).all()
+
+    def get_product(self, product_id: UUID) -> Product:
+        product = self.db.query(Product).filter(Product.id == product_id).first()
+        if not product:
+            raise HTTPException(
+                status_code=status.HTTP_404_NOT_FOUND,
+                detail=f"Product with ID {product_id} not found.",
+            )
+        return product
+
+    def search_products(self, search_request: SearchRequest):
+        query = self.db.query(Product)
+
+        if search_request.name:
+            query = query.filter(Product.name.ilike(f"%{search_request.name}%"))
+        if search_request.min_price:
+            query = query.filter(Product.price >= search_request.min_price)
+        if search_request.max_price:
+            query = query.filter(Product.price <= search_request.max_price)
+        if search_request.isAvailable is not None:
+            query = query.filter(Product.isAvailable == search_request.isAvailable)
+
+        # Sort
+        if search_request.sort_by == "price":
```

Comment:
Your search request lacks input validation. All of this could be done in a single line if asc and name are defined as the default value.

Reply:
you can look at [this](https://fastapi.tiangolo.com/tutorial/query-param-models/#query-parameters-with-a-pydantic-model) example to see how we can use the model class for query parameters

Reply:
done